---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r setup}
#set up path 
  knitr::opts_knit$set(root.dir = normalizePath("/Users/mishafaruki/Desktop/QBIO490/qbio_490_mishafaruki/analysis_data")) 
```


```{r}
#import libraries 
library(BiocManager)
library(TCGAbiolinks)
library(maftools)

#create a query based on the clinical category of TCGA-SKCM 
clin_query <- GDCquery(project = "TCGA-SKCM",
  data.category = "Clinical",
  data.type = "Clinical Supplement",
  data.format = 'BCR Biotab')
  GDCdownload(clin_query)
  clinical.BCRtab.all <- GDCprepare(clin_query)

#create a query based on the Transcriptome Profiling/RNA category of TCGA-SKCM 
rna_query <- GDCquery(project ="TCGA-SKCM",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

GDCdownload(rna_query)
#rna summarized experiment 
rna_se <- GDCprepare(rna_query)
```

1. Difference in survival between metastatic and non-metastatic patients (KM plot)
```{r}
# load required packages for KM plot 
if (!require(survival)) {
  install.packages("survival")
}

if (!require(survminer)) {
  install.packages("survminer")
}

if (!require(ggplot2)) {
  install.packages("ggplot2")
}

library(survival)
library(survminer)
library(ggplot2)

#create rna clinical that will contain data from colData
rna_clinical = rna_se@colData
#convert it to a dataframe 
rna_clinical <- as.data.frame(rna_clinical)
#Set output to Tumor_Sample_Barcode in rna_clincal so that it derived from patient category 
rna_clinical$Tumor_Sample_Barcode <- rna_clinical$patient

#Set output to vital_status_num have to contain vital_status colum's data but make the values numeric 
rna_clinical$vital_status_num <- as.numeric(rna_clinical$vital_status == "Dead")

#Create Survminer Object
#contains the survival data 
survival_object <- Surv(time = rna_clinical$days_to_death, event = rna_clinical$vital_status_num)
#fit survival_object curve 
fit <- survfit(surv_obj ~ rna_clinical$definition, data = rna_clinical)

#create KM plot 
ggsurvplot(fit, data = rna_clinical, pval = TRUE, 
           risk.table = TRUE, #risk table
           legend.title = "Metastatic Status",
           xlab = "Time", ylab = "Survival probability",
           break.time.by = 1000, # adjusting time intervals
           risk.table.col = "strata") #color risk table
```


2. Mutation differences between metastatic and non-metastatic patients for multiple genes in a Co-oncoplot.
```{r}
#access the drug and radiation dataframes
rad <- clinical.BCRtab.all$clinical_radiation_skcm[-c(1,2),]
drug <- clinical.BCRtab.all$clinical_drug_skcm[-c(1,2),]
```

```{r}
#packages
library(maftools)
```


```{r}
#instantiate object
rna_query <- GDCquery(project ="TCGA-SKCM",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)

```



```{r}
#set up maf object 
maf_object <- read.csv("/Users/mishafaruki/Desktop/QBIO490/qbio_490_mishafaruki/analysis_data/skcm_clinical_data.csv")
colnames(clinical)[ colnames(clinical) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"
write.csv(clinical, "/Users/mishafaruki/Desktop/QBIO490/qbio_490_mishafaruki/analysis_data/skcm_clinical_data.csv", row.names = FALSE)

maf_query <- GDCquery(
    project = "TCGA-SKCM", 
    data.category = "Simple Nucleotide Variation", 
    access = "open", # we only have access to somatic mutations which are open access
    data.type = "Masked Somatic Mutation", 
    workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking")

# GDCdownload(maf_query) # you shouldn't have to redownload any data!

maf <- GDCprepare(maf_query) #as long as it runs, ignore any errors

maf_object <- read.maf(maf = maf, 
                       clinicalData = rna_clinical,
                       isTCGA = TRUE)
```

```{r}
#hold clinical data of maf object 
maf_clinical = maf_object@clinical.data
maf_clinical
#Extract data in definition column that is metastatic and put it into meta 
meta <- maf_clinical[maf_clinical$definition == "Metastatic", ]
#Extract values from Tumor_Sample_Barcode of meta too and store it in there
meta <- meta$Tumor_Sample_Barcode

#Extract data in definition column that is Primary solid Tumor and put it into nonmeta 
nonmeta <- maf_clinical[maf_clinical$definition == "Primary solid Tumor", ]
#Extract values from Tumor_Sample_Barcode of nonmeta too and store it in there
nonmeta <- nonmeta$Tumor_Sample_Barcode

# Subsetting based on clinical variables, meta and nonmeta 
meta_maf <- subsetMaf(maf = maf_object, tsb = meta)
nonmeta_maf <- subsetMaf(maf = maf_object, tsb = nonmeta)

#create co-oncoplot 
coOncoplot(m1 = meta_maf,
           m2 = nonmeta_maf,
           m1Name = "Metastatic",
           m2Name = "Non-metastatic",
           borderCol = NA)
```
3. Mutation differences for specific gene of interest (one of most mutated genes)(Co-lollipop plot)
```{r}
#lollipop chart 
lollipopPlot2(m1 = meta_maf,
              m2 = nonmeta_maf,
              m1_name = "Metastatic", #why can't this be m1Name? 
              m2_name = "Non-metastatic",
              gene="TTN")
#single lollipop plot for meta 
lollipopPlot(maf=meta_maf, 
             gene="TTN")

#single lollipop plot for nonmeta 
lollipopPlot(maf=nonmeta_maf, 
             gene="TTN")
```

4. Coocurrence or mutual exclusion of common gene mutations: one for metastatic patients, one for non-metastatic patients (2 somatic interation plots)

```{r}
somatic_interaction_metastatic <- somaticInteractions(maf = meta_maf, top = 20)
somatic_interaction_nonmetastatic <- somaticInteractions(maf = nonmeta_maf, top = 20)
```

5. Differential expression between non-metastatic and metastatic patients controlling for treatment effects, race, gender, and vital status (DESeq2 + Volcano plot)
  a. Treatments must include radiation, chemotherapy, immunotherapy, molecular therapy, vaccine
  b. Treatments must include radiation, chemotherapy, immunotherapy, molecular therapy, vaccine

```{r}
#import packages 
library(SummarizedExperiment)
BiocManager::install("DESeq2")
library(DESeq2)
library(ggplot2)
library(ggrepel)
BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)
```

```{r}
chemoData <- drug[drug$pharmaceutical_therapy_type == 'Chemotherapy',]
rna_clinical$chemotherapy <- ifelse(rna_clinical$patient %in% chemoData$bcr_patient_barcode, 1, 0)

# immunotherapy
immunoData <- drug[drug$pharmaceutical_therapy_type == 'Immunotherapy',]
rna_clinical$immunotherapy <- ifelse(rna_clinical$patient %in% immunoData$bcr_patient_barcode, 1, 0)

# radiation
radiationData <- drug[drug$pharmaceutical_therapy_type == 'Ancillary',]
rna_clinical$radiation <- ifelse(rna_clinical$patient %in% radiationData$bcr_patient_barcode, 1, 0)

# molecular
molecularData <- drug[drug$pharmaceutical_therapy_type == 'Targeted Molecular therapy',]
rna_clinical$molecular <- ifelse(rna_clinical$patient %in% molecularData$bcr_patient_barcode, 1, 0)

# vaccine
vaccineData <- drug[drug$pharmaceutical_therapy_type == 'Vaccine',]
rna_clinical$vaccine <- ifelse(rna_clinical$patient %in% vaccineData$bcr_patient_barcode, 1, 0)




# Factorize all columns
rna_clinical$chemotherapy <- factor(rna_clinical$chemotherapy)
rna_clinical$immunotherapy <- factor(rna_clinical$immunotherapy)
rna_clinical$radiation <- factor(rna_clinical$radiation)
rna_clinical$molecular <- factor(rna_clinical$molecular)
rna_clinical$vaccine <- factor(rna_clinical$vaccine)

rna_clinical$race <- factor(rna_clinical$race)
rna_clinical$gender <- factor(rna_clinical$gender)
rna_clinical$vital_status <- factor(rna_clinical$vital_status)

# find counts of rna_se
countData <- assay(rna_se)

# create the DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = countData, colData = rna_clinical,
                              design = ~ chemotherapy + immunotherapy + radiation + molecular + vaccine + race + gender +vital_status)

# Filter out genes with low expression
dds <- dds[rowSums(counts(dds)) > 20, ]

# Run DESeq
dds <- DESeq(dds)

# Results with thresholding
res <- results(dds, alpha = 0.05)
res <- res[!is.na(res$padj) & abs(res$log2FoldChange) > 1, ]

# convert to data frame in order to pass to ggplot
res_df <- as.data.frame(res) 

# volcano plot using the data frame
ggplot(res_df, aes(x=log2FoldChange, y=-log10(padj))) + 
  geom_point(aes(color = padj < 0.05)) + 
  scale_color_manual(values = c("TRUE" = "red", "FALSE" = "black")) + 
  theme_minimal() + 
  labs(x = "Log2 Fold Change", y = "-Log10 adjusted p-value", title = "Volcano plot of differential expression")
```










Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

